File: src/ipc/messages/types/structs/ScanProgress.ts
import * as ldm from '../../leapdatamodel';
import * as valuetype from '../../valuetype';
import type * as types from '..';
import { StructBase as _StructBase } from '../../StructBase';
import { LeapError as _LeapError } from '../../LeapError';
import { ValueBase as _ValueBase } from '../../ValueBase';

export class ScanProgress extends _StructBase {
    static _type: string = 'scan-progress';
    static _leapStruct: valuetype.LeapStruct = new valuetype.LeapStruct("scan-progress", [], [new valuetype.Property("done-count", new valuetype.ValueType("int", []))]);
    doneCount: number;

    constructor(doneCount: number, ) {
        super();
        this.doneCount = doneCount;
    }

    _verifyAsTypeOrRaise(path: string[], valueType: valuetype.ValueType): void {
        if (valueType.name == ScanProgress._type) {
            const structType = ScanProgress._leapStruct.applyArgs(valueType.args);
            _ValueBase._verifyValueOrRaise(path.concat(['doneCount']), this.doneCount, structType.props[0].propType);
        } else {
            throw _ValueBase._pathError(path, `expecting ${ScanProgress._type}`);
        }
    }

    _toLdm(valueType: valuetype.ValueType): ldm.Value {
        let value = new Map<string, ldm.Value>();
        const structType = ScanProgress._leapStruct.applyArgs(valueType.args);
        value.set("done-count", _ValueBase._valueToLdm(this.doneCount, structType.props[0].propType));
        return new ldm.StructValue(value, valueType);
    }
}

File: src-tauri/src/ipc/messages/types/structs/scan_progress.rs
use std::collections::HashMap as _HashMap;
#[allow(unused_imports)]
use super::super::super::types;
use super::super::super::value_base::ValueBase as _ValueBase;
use super::super::super::struct_base::StructBase as _StructBase;
use super::super::super::leap_data_model as ldm;
use super::super::super::value_type::ValueType as _ValueType;

#[derive(Clone, Debug)]
pub struct ScanProgress {
    pub done_count: i64,
}

impl ScanProgress {
    pub fn new(done_count: i64, ) -> Self {
        Self {
            done_count,
        }
    }
}

impl _ValueBase for ScanProgress {
    const _TYPE: &'static str = "scan-progress";

    fn _applied_args() -> Vec<_ValueType> {
        vec![
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Struct(self._to_ldm_struct())
    }
}

impl _StructBase for ScanProgress {
    fn _to_ldm_struct(&self) -> ldm::StructValue {
        let mut value = _HashMap::new();
        value.insert("done-count".to_owned(), self.done_count._to_ldm());
        ldm::StructValue{
            value,
            value_type: _ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        }
    }
}